src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs

```diff
@@ -62,34 +72,50 @@ public bool IsStarted
         /// <param name="elapsed">The task to run.</param>
         public void Start(Func<Task> elapsed)
         {
-            var newTokenSource = new CancellationTokenSource();
+            DelayedWork delayedWork = new DelayedWork(elapsed);
+            DelayedWork previousWork = Interlocked.Exchange(ref this.latestWork, delayedWork);
+
+            if (previousWork != null)
+            {
+                previousWork.IsCancellationRequested = true;
+            }
 
-            TaskEx.Delay(this.Delay, newTokenSource.Token)
-                .ContinueWith(
 #if !NET40
-                async previousTask =>
+            Task delayTask = Task.Delay(this.Delay);
 #else
-                    previousTask =>
+            Task delayTask = TaskEx.Delay(this.Delay);
 #endif
+            delayTask.ContinueWith(
+                    (dlyTask) =>
                     {
-                        CancelAndDispose(Interlocked.CompareExchange(ref this.tokenSource, null, newTokenSource));
+                        Interlocked.CompareExchange(ref this.latestWork, null, delayedWork);
+                        if (delayedWork.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
                         try
                         {
-                            Task task = elapsed();
+                            // Task may (or may not) be executed syncronously
+                            Task task = (delayedWork.RunFunc == null)
+                                                ? null
+                                                : delayedWork.RunFunc();
 
-                            // Task may be executed syncronously
-                            // It should return Task.FromResult but just in case we check for null if someone returned null
-                            if (task != null)
+                            // Just in case we check for null if someone returned null
+                            if (task == null)
                             {
-#if !NET40
-                                await task.ConfigureAwait(false);
-#else
-                                task.ContinueWith(
-                                    userTask =>
+                                return;
+                            }
+
+                            task.ContinueWith(
+                                    (userTask) =>
                                     {
                                         try
                                         {
-                                            userTask.RethrowIfFaulted();
+                                            if (task.IsFaulted)
+                                            {
+                                                throw new AggregateException(task.Exception).Flatten();
```

Comment:
Why do we need to wrap then flatten the original exception here when `LogException` already does it?